### ISSUE TITLE
question: computedFK with relation?

### ISSUE THREAD

**Issue by unknown1337** [supabase/pg_graphql#236]
Hi all,

**question**
is it possible to create a relation between two tables with a **computed** column as FK for pg_graphql? Or a workaround?

**summary** 
computed column returns FKs based on several filter criteria in other tables. Though there is no resulting collection to query the corresponding table elements

**Situation**
tried my best, please let me know if things are not clear :)

1. relations: `groups`<->`filters` and through the filters `group`<->`cars`. M2M.
- table: `groups` tree (columns: groupName, parentGroupId) & ( groups_filters join table)
- table: `filters` (columns: groupId, field, operator, value, groupId).  Ex{1,'color','EQUAL','red'}
2. UI: allows user to create a tree of filter `groups`. User is then able to (for a given `group`) get all `cars `based on all `filters `belonging to that `group `and the parent `groups `. 
- Ex: if group 1 has a filter that country equals netherlands and the parent group 0 has a filter that the color equals red. Then group one should display all red cars from the netherlands 
3. Query: to implement point 2:
- I created in Supabase a  plpg `GetCarIdsForGroupId `function that first recursively loads all `filters `(incl parents) belonging to a given and its parents `group`. Then the function creates a query string based on all those `filter `fields, operators, values and executes. Function returns carIds 
- I've created a computed field `CarIds `on table `groups `according the [doc ](https://supabase.github.io/pg_graphql/configuration/#computed-field) that execute the created function with the rec.groupId as arg

All of the above works :), but in order to get to the `cars ` (incl its relations) for a given group I have to execute two graqphqlqueries
1. query getCarIdsForGroupId...groupsCollection(filter:{id:{eq:$groupId}}){edge{node{groupName, `carIds`}}} 
2. query getCarsForCarIds....carsCollection(filter:{id:{in:$carIds}}){edge{node{brand, color, plate,...}}}

**while I would like to do this in one graphql query** without hardcoding all fields in the   `GetCarIdsForGroupId ` function

THANKS!

**Comment by olirice** [comment]
If I understand correctly, the end result of what you have is something like

```sql
create table group(
  id int primary key,
  car_ids int[],         -- computed from function, references car(id)
)

create table car(
  id int primary key,
)
```
where elements in `group.car_ids` represent an unenforced foreign key reference to `car.id`.

----

Postgres does not allow enforcement of array foreign keys, instead, you could normalize it out into a separate table

```sql
create table group(
  id int primary key
);

create table group_to_cars(
  id int primary key,
  group_id int not null references group(id),
  car_id int not null references car(id)
);

create table car(
  id int primary key,
);
```

where you'd use a trigger function/s to keep `group_to_cars` up to date and could then retrieve all the data you wanted in a single query via

```graphql
query getCarsForCarIds {
  groupsCollection {
    edges {
      node {
        id
        groupToCarsCollection {
          edges {
            node {
              carsCollection {
                edges {
                  node {
                    id
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

```

**Comment by unknown1337** [comment]
Thanks a lot @olirice for your clear reply!

1. your suggestion: Thanks for the suggestion, did not think of it like that! (This also brings the additional benefit of 'caching' the applied filters result in group_to_cars iso the 'costly recursive queries'). I will keep this idea for later!
2. other: am I correct that without creating additional sql tables or hardcoding the fields in `GetCarIdsForGroupId ` or performing two consequtive graphql queries (1. getIds. 2. getCarsCollection)  this problem cannot be solved. As e.g. graphql does not support 'nested' queries with some intermediate `myTmpvar`
```
query getCarsForCarIds {
  groupsCollection {
    edges {
      node {
        id
        carIds into myTmpvar
        --->query.carsCollection(filter: {id:{in: myTmpvar}}) { <----
                 edges {
                    node {
                      id
                     }
                  }
                 }
```

**Comment by olirice** [comment]
> am I correct that without creating additional sql tables or hardcoding the fields in GetCarIdsForGroupId or performing two consequtive graphql queries (1. getIds. 2. getCarsCollection) this problem cannot be solved. As e.g. graphql does not support 'nested' queries with some intermediate myTmpvar

correct, you can not use a value from earlier in the query as a variable in later parts of the query

--

> the philosophy to never store data twice in a db as this needs to be kept in sync

Once we add support for views there will be a better solution to this kind of computed relationship problem because the `group_to_cars` table could be a view (so you can compute the relationship on the fly). Support for views will be coming in the next few months
